infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java

```diff
@@ -54,83 +52,153 @@ public LogEntryWriter(CorfuRuntime rt, LogReplicationConfig config, LogReplicati
+                            log.warn("Message metadata mismatch. Skip applying message {}, persistedTopologyConfigId={}," +
+                                    "persistedSnapshotStart={}, persistedSnapshotDone={}, persistedBatchTs={}",
+                                txMessage.getMetadata(), persistedTopologyConfigId, persistedSnapshotStart,
+                                persistedSnapshotDone, persistedBatchTs);
+                            throw new IllegalArgumentException("Cannot apply log entry message due to metadata mismatch");
+                        }
+
+                        // Skip Opaque entries with timestamp that are not larger than persistedOpaqueEntryTs
+                        if (opaqueEntry.getVersion() <= persistedOpaqueEntryTs) {
+                            log.trace("Skipping entry {} as it is less than the last applied opaque entry {}",
+                                opaqueEntry.getVersion(), persistedOpaqueEntryTs);
+                            return null;
+                        }
+
+                        // LAST_LOG_ENTRY_APPLIED has the timestamp of the last OpaqueEntry applied from a
+                        // batch of opaque entries received.
+                        logReplicationMetadataManager.appendUpdate(txnContext,
+                            LogReplicationMetadataType.LAST_LOG_ENTRY_APPLIED, opaqueEntry.getVersion());
+
+                        // CorfuStore uses WriteAfterWriteTransaction type so even though the topology is read
+                        // and validated for each OpaqueEntry, this read will not be used to detect a concurrent
+                        // topology update on the Sink.  So force an update to this key by using the touch() api.
+                        // NOTE: This will be addressed once the schema of the metadata table is updated to have
+                        // all types in a single key (remote session)
+                        logReplicationMetadataManager.touch(txnContext, LogReplicationMetadataType.TOPOLOGY_CONFIG_ID);
+
+                        // If this is the last OpaqueEntry in the message/batch, update LAST_LOG_ENTRY_BATCH_PROCESSED
+                        // with its timestamp
+                        if (opaqueEntry.getVersion() == txMessage.getMetadata().getTimestamp()) {
+                            logReplicationMetadataManager.appendUpdate(txnContext,
+                                LogReplicationMetadataType.LAST_LOG_ENTRY_BATCH_PROCESSED,
+                                txMessage.getMetadata().getTimestamp());
+                        }
+
+                        for (UUID streamId : opaqueEntry.getEntries().keySet()) {
+                            if (ignoreEntriesForStream(streamId)) {
+                                log.warn("Skip applying log entries for stream {} as it is noisy. The Source and" +
+                                    "Sink sites could be on different versions", streamId);
+                                continue;
```

Comment:
continue statement will not work here as this is in an IRetry block.  You will need a return statement to move to the next opaque entry.  See L125 for reference.